src/server/experiments.rs

```diff
@@ -0,0 +1,461 @@
+        self.experiment.mode = mode;
+        Ok(())
+    }
+
+    pub fn set_cap_lints(&mut self, db: &Database, cap_lints: ExCapLints) -> Result<()> {
+        db.execute(
+            "UPDATE experiments SET cap_lints = ?1 WHERE name = ?2;",
+            &[&cap_lints.to_str(), &self.experiment.name.as_str()],
+        )?;
+        self.experiment.cap_lints = cap_lints;
+        Ok(())
+    }
+
+    pub fn set_priority(&mut self, db: &Database, priority: i32) -> Result<()> {
+        db.execute(
+            "UPDATE experiments SET priority = ?1 WHERE name = ?2;",
+            &[&priority, &self.experiment.name.as_str()],
+        )?;
+        self.server_data.priority = priority;
+        Ok(())
+    }
+
+    pub fn set_crates(&mut self, db: &Database, crates: Vec<Crate>) -> Result<()> {
+        db.transaction(|transaction| {
+            transaction.execute(
+                "DELETE FROM experiment_crates WHERE experiment = ?1;",
+                &[&self.experiment.name.as_str()],
+            )?;
+
+            for krate in &crates {
+                transaction.execute(
+                    "INSERT INTO experiment_crates (experiment, crate) VALUES (?1, ?2);",
+                    &[
+                        &self.experiment.name.as_str(),
+                        &serde_json::to_string(&krate)?,
+                    ],
+                )?;
+            }
+
+            Ok(())
+        })?;
+        self.experiment.crates = crates;
+        Ok(())
+    }
+
+    pub fn set_start_toolchain(&mut self, db: &Database, start: Toolchain) -> Result<()> {
+        db.execute(
+            "UPDATE experiments SET toolchain_start = ?1 WHERE name = ?2;",
+            &[
+                &serde_json::to_string(&start)?,
+                &self.experiment.name.as_str(),
+            ],
+        )?;
+        self.experiment.toolchains[0] = start;
+        Ok(())
+    }
+
+    pub fn set_end_toolchain(&mut self, db: &Database, end: Toolchain) -> Result<()> {
+        db.execute(
+            "UPDATE experiments SET toolchain_end = ?1 WHERE name = ?2;",
+            &[
+                &serde_json::to_string(&end)?,
+                &self.experiment.name.as_str(),
+            ],
+        )?;
+        self.experiment.toolchains[1] = end;
+        Ok(())
+    }
+}
+
+struct ExperimentDBRecord {
+    name: String,
+    mode: String,
+    cap_lints: String,
+    toolchain_start: String,
+    toolchain_end: String,
+    priority: i32,
+    created_at: DateTime<Utc>,
+    github_issue: Option<String>,
+    github_issue_url: Option<String>,
+    github_issue_number: Option<i32>,
+    status: String,
+    assigned_to: Option<String>,
+}
+
+impl ExperimentDBRecord {
+    fn from_row(row: &Row) -> Self {
+        ExperimentDBRecord {
+            name: row.get("name"),
+            mode: row.get("mode"),
+            cap_lints: row.get("cap_lints"),
+            toolchain_start: row.get("toolchain_start"),
+            toolchain_end: row.get("toolchain_end"),
```

Comment:
`serde_json::from_str` for these two lines?

Reply:
Uhm, I'm not sure this is the right approach. I had issues with the borrowchecker, so I had to split the conversion from a database row to an experiment with that intermediate struct. Now that there is this split (`from_row` extracts the data from the database and `into_experiment_data` converts it to the proper structs) why move things around?